### ISSUE TITLE
Rewritten signature generation for a struct

### ISSUE THREAD

**Issue by buybackoff** [mono/CppSharp#1324]
##### Brief Description

When generating signature for an *original* definition from OpenBLAS distribution:

```
  typedef struct { double real, imag; } openblas_complex_double;
```
and function 

```
openblas_complex_double cblas_zdotc(OPENBLAS_CONST blasint n, OPENBLAS_CONST void *x, OPENBLAS_CONST blasint incx, OPENBLAS_CONST void *y, OPENBLAS_CONST blasint incy);
```

The generated method is:

```
[SuppressUnmanagedCodeSecurity]
            [DllImport("OpenBLAS", CallingConvention = global::System.Runtime.InteropServices.CallingConvention.Cdecl,
                EntryPoint="cblas_zdotc")]
            internal static extern void CblasZdotc(global::System.IntPtr @return, int n, global::System.IntPtr x, int incx, global::System.IntPtr y, int incy);

```
I.e. the return value is placed as the first argument.

If I change the definition to a *wrong* one (replace double with float ):
```
  typedef struct { float real, imag; } openblas_complex_double;
```
the generated output is OK:

```
[SuppressUnmanagedCodeSecurity]
            [DllImport("OpenBLAS", CallingConvention = global::System.Runtime.InteropServices.CallingConvention.Cdecl,
                EntryPoint="cblas_zdotc")]
            internal static extern global::BLAS.OpenBLAS.OpenblasComplexDouble.__Internal CblasZdotc(int n, global::System.IntPtr x, int incx, global::System.IntPtr y, int incy);

```

For me it's OK to have wrong headers just for codegen since I replace the structs with C#-defined ones via TypeMap and only need a distinc name, but the behavior looks wrong.

OS: Windows 10, latest NuGet binaries

##### Used settings
```
public void Setup(Driver driver)
        {
            var options = driver.Options;
            options.GeneratorKind = GeneratorKind.CSharp;

            options.OutputDir = @"..\..\Generated\OpenBLAS";
            options.Verbose = true;
            options.GenerateSingleCSharpFile = false;
            options.GenerateFunctionTemplates = false;
            
            driver.Options.CheckSymbols = false;

            var module = options.AddModule("OpenBLAS");

            module.IncludeDirs.Add(@"..\..\..\..\lib\OpenBLAS\include\");

            module.Headers.Add("cblas.h");

            module.Defines.Add("HAVE_LAPACK_CONFIG_H");
            module.Defines.Add("LAPACK_COMPLEX_STRUCTURE");
            module.Defines.Add("OPENBLAS_COMPLEX_STRUCT");
        }

public void Preprocess(Driver driver, ASTContext ctx)
        {
            // uncommenting doesn't affect the behavior
            // driver.Context.TypeMaps.TypeMaps.Add("openblas_complex_float", new OpenBlasSingle());
            // driver.Context.TypeMaps.TypeMaps.Add("openblas_complex_double", new OpenBlasDouble());
            ctx.IgnoreFunctionWithPattern("^/(?!LAPACKE|openblas|cblas)([A-Za-z0-9_]+)$");
        }
```
Target: MSVC

**Comment by tritao** [comment]
Have you verified at runtime that the generated is wrong, i.e. crashes?

https://godbolt.org/z/2eRgwY

This simple repro in Godbolt confirms that the generated code looks correct for both 32-bit and 64-bit MSVC ABI (you change triple to `x86_64-windows-msvc` to confirm yourself).

**Comment by buybackoff** [comment]
> Have you verified at runtime that the generated code is wrong, i.e. crashes?

Not yet, I'm fighting with CppSharp for the good part of the day just to get consistent DllImports without simple structs being replaced by 100+-lines stubs, and only then could test PInvokes.

Even if it's correct, why the signature is weird vs it's float counterpart, which is generated normally? How a return value of s simple struct could be replaced by void and instead a void pointer is provided as the first argument? If this is how C works I do not know/get it.

**Comment by tritao** [comment]
> Not yet, I'm fighting with CppSharp for the good part of the day just to get consistent DllImports without simple structs being replaced by 100+-lines stubs, and only then could test PInvokes.

Not really sure what you mean by consistency here, but it sounds like you'd prefer a more streamlined approach to binding, with simpler C# structs and P/Invokes layer that you can build upon on the C# side? 

> Even if it's correct, why the signature is weird vs it's float counterpart, which is generated normally? How a return value of s simple struct could be replaced by void and instead a void pointer is provided as the first argument? If this is how C works I do not know/get it.

It is correct, the C ABI can differ from the C API in user code. 

Returns of structs by value are a notorious part of this.

The reason that the float version returns directly is that it ends up being 64-bit sized, which can fit in a register directly. For the double version it exceeds the register size, and so the compiler decides to return it indirectly. 

This is usually transparent to you when using the C API with a C compiler, but absolutely important to get right when invoking such C APIs directly.

**Comment by buybackoff** [comment]
> Not really sure what you mean by consistency here

I just want PInvoke signatures to look like in C headers and the documentation.

> a more streamlined approach to binding, with simpler C# structs and P/Invokes layer that you can build upon on the C# side?

You know, BLAS/LAPACKE have hundreads of functions, doing that manually is not possible, SharpGenTools doesn't work. CppSharp now works fine for me, quarantine helped to fight with it a little :)

Thanks for explanation, I thought that something was going on here beyond my understanding.

Will .NET do correct marshalling for me if I just keep it as below, with BLAS.ComplexDouble being a 16-bytes struct defined in C#?

```
internal static extern BLAS.ComplexDouble CblasZdotc(int n, ...)
```

**Comment by tritao** [comment]
> I just want PInvoke signatures to look like in C headers and the documentation.
> 
> >

Then use the higher-level layer generated by CppSharp which should solve the mismatch for you.

> You know, BLAS/LAPACKE have hundreads of functions, doing that manually is not possible, SharpGenTools doesn't work. CppSharp now works fine for me, quarantine helped to fight with it a little :)

What were the main problems you had with CppSharp? We definitely welcome any feedback.

> Will .NET do correct marshalling for me if I just keep it as below, with BLAS.ComplexDouble being a 16-bytes struct defined in C#?
> 
> ```
> internal static extern BLAS.ComplexDouble CblasZdotc(int n, ...)
> ```

I don't think that will work, I recommend you to use the signatures generated by CppSharp or you will have problems due to mismatches.

**Comment by buybackoff** [comment]
> What were the main problems you had with CppSharp? We definitely welcome any feedback.

Simple options switch to disable higher-level layer and a simple way to replace a C type with existing C#-defined type (at least blittable struct) are needed.

The main problem is that higher-level abstraction is not useful in this case.

> Then use the higher-level layer generated by CppSharp which should solve the mismatch for you.

I would like an option to use raw pointers, without deeply-nested `fixed` blocks that pin references to C# primitives, which in my particular case are already guaranteed to be pinned. Also, the generated code for DllImports requires an assembly to be unsafe, there is no point of hiding unsafe pointers behind "safe" higher-level API. That should be responsibility of user code to pin what is needed to be pinned.

I had to read a lot of issues, to google, to read code from links to projects that use CppSharp (those links are the best part of limited docs) to learn how to replace a native type with C#-defined type. And I still couldn't find a way to use a pointer to C#-defined blittable type as input to a native function, it's just replaced with IntPtr. Well, IntPtr is OK, but replacement should be simple.

I will probably never use the functions that caused the most troubles here, and in retrospect I would had spent less time just by manually defining only the DllImports that I needed. But a tool to generate bindings is very useful and often needed just to save time from manual tedious error-prone typing. Just for DllImports, without forcing higher-level layer that makes things slower and more complex. I used SharpGenTools recently for a different library, but it fails badly with this one, and CppSharp looks better overall, yet more complex for simple customization.

And finally, small annoyance: NuGet did not work (on Rider with VS's MSBuild) - it doesn't copy output folder, I had to do so manually. The dependency `Baseclass.Contrib.Nuget.Output` doesn't do its work. It would be better to include `.targets` file directly into CppSharp NuGet.
